llvm/lib/Analysis/CmpInstAnalysis.cpp

```diff
@@ -176,6 +177,24 @@ llvm::decomposeBitTest(Value *Cond, bool LookThruTrunc, bool AllowNonZeroC) {
                                 ICmp->getPredicate(), LookThruTrunc,
                                 AllowNonZeroC);
   }
+  Value *X;
+  if (Cond->getType()->isIntOrIntVectorTy(1) &&
+      (match(Cond, m_Trunc(m_Value(X))) ||
+       match(Cond, m_Not(m_Trunc(m_Value(X)))))) {
+    DecomposedBitTest Result;
+    Result.X = X;
+    unsigned BitWidth = X->getType()->getScalarSizeInBits();
+    Result.Mask = APInt(BitWidth, 1);
+    Result.C = APInt::getZero(BitWidth);
+    Result.Pred = isa<TruncInst>(Cond) ? ICmpInst::ICMP_NE : ICmpInst::ICMP_EQ;
+
+    if (LookThruTrunc && match(Result.X, m_Trunc(m_Value(X)))) {
+      Result.X = X;
+      Result.Mask = Result.Mask.zext(X->getType()->getScalarSizeInBits());
+      Result.C = Result.C.zext(X->getType()->getScalarSizeInBits());
+    }
```

Comment:
Think the `LookThruTrunc` logic is untested.

Reply:
What is the motivation for `LookThruTrunc`? Doesn't `(trunc (trunc X))` always get folded to `(trunc X)` in InstCombine?

Reply:
There seems to be no motivation to handle the LookThruTrunc for the trunc case as you say it is folded to one trunc, I did not think all the way and thought that an extra use of the first trunc stopped that fold but it did not.